Oracle oci builder

SwampDragons opened the pull request:
MVP Oracle Classic Builder.  Comments welcome.

mwhooker commented on builder/amazon/common/step_key_pair.go:
```diff
@@ -36,7 +37,7 @@ func (s *StepKeyPair) Run(_ context.Context, state multistep.StateBag) multistep
 
 		state.Put("keyPair", s.KeyPairName)
-		state.Put("privateKey", string(privateKeyBytes))
+		state.Put("privateKey", strings.TrimSpace(string(privateKeyBytes)))
```
is this change supposed to be here?

SwampDragons replied:
I must have been doing some testing of a different issue on this branch. removed it.

mwhooker commented on builder/oracle/classic/config.go:
```diff
@@ -0,0 +1,67 @@
+	common.PackerConfig `mapstructure:",squash"`
+	Comm                communicator.Config `mapstructure:",squash"`
+
+	Access *AccessConfig
```
should either move access config to this struct or remove `AccessConfig`.

SwampDragons replied:
Pulled it. I don't think we were using it.

mwhooker commented on builder/oracle/classic/builder.go:
```diff
@@ -0,0 +1,107 @@
+	if err != nil {
+		return nil, err
+	}
+	b.config = config
```
validate that required config is present

SwampDragons replied:
Should I do this here or inside of the NewConfig function?

mwhooker replied:
inside NewConfig, since I think it's good to keep it all together.

Marcus-James-Adams replied:
will this now support sending a --user-data-file?in the same way, as you can do so in AWS?  It is now a feature of the oci command line program?   If you are using the standard Oracle windows severe 2012 images you need to be able to send a "setup" Powershell /batch file to make sure the windows rm and firewalls are configured correctly so that packer can build, especially when packer may be in one Oracle datacenter and you are building in another?

SwampDragons replied:
This is an MVP of the builder and does not yet have that option. Would you please open a new issue with your feature request?

mwhooker commented on builder/oracle/classic/builder.go:
```diff
@@ -0,0 +1,107 @@
+
+	httpClient := cleanhttp.DefaultClient()
+	config := &opc.Config{
+		Username:       opc.String(b.config.Username),
```
should we get config from the environment? Wondering if we should mirror TF here

mwhooker commented on builder/oracle/classic/step_add_keys.go:
```diff
@@ -0,0 +1,58 @@
+
+	// Load the packer-generated SSH key into the Oracle Compute cloud.
+	keyInfo, err := sshKeysClient.CreateSSHKey(&sshKeysInput)
+	if err != nil {
```
I think we should fail if a key exists. I might be missing some context around keys, but in general we should be creating keys with collision-resistant names (i.e. uuids). The other use-case is where the user has a key they want to use, and will pass in the name directly. We should make sure we support both of these if it's technically possible.

SwampDragons replied:
That's fine. I'll change this to use a UUID instead of a static name.

mwhooker commented on builder/oracle/classic/step_add_keys.go:
```diff
@@ -0,0 +1,58 @@
+				Key:     sshPublicKey,
+				Enabled: true,
+			}
+			keyInfo, err = sshKeysClient.UpdateSSHKey(&updateKeysInput)
```
we don't check the error here, but I think this will be obviated by the comment above

mwhooker commented on builder/oracle/classic/step_add_keys.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+func (s *stepAddKeysToAPI) Cleanup(state multistep.StateBag) {
+	// Nothing to do
```
we should delete keys we created here

SwampDragons replied:
👍

mwhooker commented on builder/oracle/classic/step_security.go:
```diff
@@ -0,0 +1,96 @@
+		config.IdentityDomain, config.Username, config.ImageName)
+	_, err = secRulesClient.CreateSecRule(&secRulesInput)
+	if err != nil {
+		log.Printf(err.Error())
```
we should wrap these in `fmt.Errorf` with some context about what's going on

mwhooker commented on builder/oracle/classic/step_create_ip_reservation.go:
```diff
@@ -0,0 +1,41 @@
+		return multistep.ActionHalt
+	}
+	state.Put("instance_ip", ipRes.IP)
+	log.Printf("debug: ipRes is %#v", ipRes)
```
is this still useful debugging info?

SwampDragons replied:
Probably not.

mwhooker commented on builder/oracle/classic/builder.go:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	// Run the steps
+	b.runner = common.NewRunner(steps, b.config.PackerConfig, ui)
```
We should probably remove `NewRunner` now that it's in `helper/multirunner`

mwhooker replied:
actually looked into this and the only difference between `NewRunner` and `NewRunnerWithPauseFn` is the latter adds the pause function to the state bag. This is needed only if you want to pause within a step, and not between step boundaries. I don't think we need to pause intra step here

mwhooker commented on builder/oracle/classic/config.go:
```diff
@@ -0,0 +1,84 @@
+
+	// Validate that all required fields are present
+	var errs *packer.MultiError
+	required := map[string]string{
```
neat

mwhooker commented on builder/oracle/classic/config.go:
```diff
@@ -0,0 +1,84 @@
+		errs = packer.MultiErrorAppend(errs, es...)
+	}
+
+	if es := c.Comm.Prepare(&c.ctx); len(es) > 0 {
```
double comm prep?

mwhooker replied:
double prep here

mwhooker commented on builder/oracle/classic/step_add_keys.go:
```diff
@@ -0,0 +1,67 @@
+		ui.Error(fmt.Sprintf("Error creating unique SSH key name: %s", err.Error()))
+		return multistep.ActionHalt
+	}
+	sshKeyName := fmt.Sprintf("/Compute-%s/%s/packer_generated_key_%s",
```
do we want to support the case where the user is bringing their own key? Gets more complicated (see `builder/amazon/common/step_key_pair.go`)...

mwhooker commented:
I think there is 1, maybe 2 more things to do and then we'll be ready for further review:

- [ ] clean up ip reservations
- [ ] specify preexisting key pair names.

the 2nd one is debatable. I'm happy to do it because I think it'll only be 5-6 lines, but testing it will be meh

scross01 commented on builder/oracle/classic/artifact.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+func (a *Artifact) Id() string {
+	return a.Snapshot.Name
```
Final Artifact should be an [ImageList Entry](https://docs.oracle.com/en/cloud/iaas/compute-iaas-cloud/stcsa/api-ImageListEntries.html) for the created machine image, not the snapshot.  The ImageList and ImageList Entry are the resources needed to enable the created image to be used for launching instances of the image via the UI or APIs. The Snapshot resource can be deleted, it doesn't need to be part of the final artifact.

For the MVP I'd suggest creating a new ImageList and ImageList Entry for each build, a enhancement would be to add a new entry to an existing image list.

SwampDragons replied:
It turns out that adding a new entry to an existing list is pretty easy; I'll go ahead and add it to this mvp since it's fresh in my mind.

scross01 commented on builder/oracle/classic/step_snapshot.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+func (s *stepSnapshot) Cleanup(state multistep.StateBag) {
+	// Nothing to do
```
The Snapshot resource can be removed during cleanup.  Note: [deleting the Snapshop](https://docs.oracle.com/en/cloud/iaas/compute-iaas-cloud/stcsa/op-snapshot-%7Bname%7D-delete.html) resource only deletes the snapshop metadata - the [MachineImage](https://docs.oracle.com/en/cloud/iaas/compute-iaas-cloud/stcsa/api-MachineImages.html) resource that created as a side effect of creating snapshot is not removed, and the actual image file in the object storage is also retained.

SwampDragons replied:
So I should delete the snapshot resource, but not the machine image?  Or is it duplicated when I add it to the image list, and I need to delete both?

scross01 replied:
Just delete the snapshot.

scross01 commented on builder/oracle/classic/builder.go:
```diff
@@ -0,0 +1,114 @@
+		},
+		&common.StepProvision{},
+		&stepSnapshot{},
+	}
```
Need a final step to create the [ImageList](https://docs.oracle.com/en/cloud/iaas/compute-iaas-cloud/stcsa/api-ImageLists.html) and [ImageListEntry](https://docs.oracle.com/en/cloud/iaas/compute-iaas-cloud/stcsa/api-ImageListEntries.html) for the machine image

scross01 commented on builder/oracle/classic/artifact.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+func (a *Artifact) String() string {
+	return fmt.Sprintf("A Snapshot was created: \n"+
```
Useful attributes to return would be:
- ImageList `name` - e.g. `/Compute-acme/user@example.com/Packer_builder_test_12345678`
- ImageList Entry `version` e.g. 1
- MachineImage `file` -  e.g. `image/8b968684-6c7a-460c-a89e-c278851a0311-8e3a0496-d35c-43b9-874c-ff96b5db4093-snapshot.tar.gz`

SwampDragons replied:
so I had a question about the ImageList entry versions -- should this be a user-required variable? Or should Packer check what is in the imagelist already and increment on top of that?

scross01 replied:
I think packer should determine the next entry to use automagically. Might be nice to provide an optional attribute to set it, but I'm not sure ()haven't personally tested) what happens if you try to reuse a version, or try to create an entry out of sequence, so picking next available is probably less error prone.

SwampDragons replied:
Yeah I checked it and the API gets unhappy if you try to set an already-used version; I've set it up to automagically increment.

scross01 commented on builder/oracle/classic/config.go:
```diff
@@ -23,9 +23,13 @@ type Config struct {
+	DestImageList   string `mapstructure:"dest_image_list"`
+	// Optional; if you don't enter anything, the image list description
+	// will read "Packer-built image list"
+	DestImageListDescription string `mapstructure:"dest_image_list_description`
```
I think this attribute could be named just `image_description` - `dest_image_list_description` seems overly long

SwampDragons commented:
@scross01 You reviewed too soon this morning!  I'm ready for you now.  :)

scross01 commented on builder/oracle/classic/step_list_images.go:
```diff
@@ -0,0 +1,92 @@
+			Description: "Packer-built image list",
+		}
+
+		// Load the packer-generated SSH key into the Oracle Compute cloud.
```
This comment seems out of place (SSH key?)

scross01 commented on builder/oracle/classic/step_list_images.go:
```diff
@@ -0,0 +1,70 @@
+	entriesClient := client.ImageListEntries()
+	entriesInput := compute.CreateImageListEntryInput{
+		Name:          config.DestImageList,
+		MachineImages: []string{fmt.Sprintf("Compute-%s/%s/%s", config.IdentityDomain, config.Username, snap.MachineImage)},
```
BUG: the machine image name needs to start with a `/Compute`, but is current setting just `Compute` so the machine iamge doesn't get associated correctly

i.e. ...`Sprintf("/Compute-%s/%s/%s",`...

scross01 commented:
One (final) issue, the `default` attribute of the image list list needs to get updated to the `version` of the new image list entry that is created.  When the image list is create the `default` is set to 0, so when the first entry is added `default` needs to be a updated to 1, etc. This ensure the latest version is used as the default when launching an instance from the image.

SwampDragons commented:
@scross01 I'm gonna hold off on merging until you give me the explicit go-ahead, but I believe I have addressed all your comments.  Thanks for being so engaged and responsive!

scross01 commented on website/source/docs/builders/oracle-classic.html.md:
```diff
@@ -0,0 +1,105 @@
+Type: `oracle-classic`
+
+The `oracle-classic` Packer builder is able to create custom images for use
+with [Oracle Compute Cloud](https://cloud.oracle.com/compute-classic). The builder
```
Office branded name "Oracle Cloud Infrastructure Classic Compute"

scross01 commented on website/source/docs/builders/oracle-classic.html.md:
```diff
@@ -0,0 +1,105 @@
+sidebar_current: 'docs-builders-oracle-classic'
+---
+
+# Oracle Compute Cloud (Classic) Builder
```
"Oracle Infrastructure Cloud Classic Compute"

scross01 commented on website/source/docs/builders/oracle-classic.html.md:
```diff
@@ -0,0 +1,105 @@
+  The oracle-classic builder is able to create new custom images for use with Oracle
+  Compute Cloud.
+layout: docs
+page_title: 'Oracle Classic - Builders'
```
Oracle Cloud Infrastructure Classic - Builders

scross01 commented:
@SwampDragons just a couple of final comments on the branding in the doc. Other than that everything looks good for MVP release.
